corehq/apps/reports/v2/models.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+from collections import namedtuple
+
+from corehq.apps.reports.v2.exceptions import EndpointNotFoundError
+
+
+EndpointContext = namedtuple('EndpointContext', 'slug urlname')
+
+
+class BaseReport(object):
+    """
+    This is the glue that ties all the moving parts of a report together.
+    """
+    slug = None
+    data_endpoints = ()
+    filter_endpoints = ()
+    formatters = ()
```

Comment:
Unrelated comment - I don't know if it's feasible, but it might be fun to try and implement this entire class with composition and disallow inheritance entirely. In that world these wouldn't be class-level attributes that get overridden, but arguments that get passed in a constructor. That really forces you to define your interfaces/components carefully.